### PR TITLE
ban: Refactor and add new option format

### PR DESCRIPTION
**Pull request by ajafff**
#### PR checklist

- [x] Addresses an existing issue: #1940, #2821
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
Refactor and simplify banRule. 
[enhancement] `ban` new options format: allows to specify an optional explanation message for function bans, banning nested methods and using a wildcard for object of a method ban
Fixes: #1940
Fixes: #2821

#### Is there anything you'd like reviewers to focus on?

This removes some ancient rather unintentional behavior, where banning `foo.bar()` would also ban `obj.foo.bar()` and even `fn().foo.bar()`.
I looked through the git history and PRs for the rule and it seems adding this "feature" was unintended.

### CODE REVIEW

File: test/rules/ban/tslint.json [hunk 1]
```diff
@@ -2,8 +2,8 @@
   "rules": {
     "ban": [
       true,
-      ["xit"],
-      ["fit"],
+      "xit",
+      {"name": "fit"},
```
Comment (nchen63) [inline test/rules/ban/tslint.json]: add test for `message` in object

File: src/rules/banRule.ts [hunk 2]
```diff
@@ -54,74 +66,66 @@ export class Rule extends Lint.Rules.AbstractRule {
...
-        } else if (bannedFunction.length >= 2) {
-            this.bannedFunctions.push(bannedFunction);
+function parseOptions(args: any[]): Options {
+    const functions: FunctionBan[] = [];
+    const methods: MethodBan[] = [];
+    for (const arg of args) {
+        if (typeof arg === "string") {
+            functions.push({name: arg});
+        } else if (Array.isArray(arg)) {
+            if (arg.length === 1) {
```
Comment (nchen63) [inline src/rules/banRule.ts]: handle arg.length == 0

File: src/rules/banRule.ts [hunk 3]
```diff
@@ -14,34 +14,67 @@
...
         optionExamples: [
             [
                 true,
-                ["someGlobalMethod"],
-                ["someObject", "someFunction"],
-                ["someObject", "otherFunction", "Optional explanation"],
+                "eval",
+                {name: "$", message: "please don't"},
+                ["describe", "only"],
+                {object: "it", name: "only", message: "don't fucus tests"},
```
Comment (nchen63) [inline src/rules/banRule.ts]: "fucus" -> "focus"?

**Comment by nchen63** [comment]
what if someone actually wanted to ban obj.foo.bar()?

**Comment by ajafff** [comment]
> what if someone actually wanted to ban obj.foo.bar()?

That's really prone to errors, because it would also ban `obj2.foo.bar()` which may not be the desired behavior. Maybe we could extend the config to take an array for the `object` property.
For the above example you could then write: 
```json
{"object": ["obj", "foo"], "name": "bar"}
```

**Comment by adidahiya** [comment]
bump @ajafff, can you address @nchen63's comments?

**Comment by ajafff** [comment]
@adidahiya I actually did that and messed up a rebase recently ... will fix that

**Comment by ajafff** [comment]
While thinking about #2821 I realized that there shouldn't be an `object` property. Instead `name` should be an array like it was before. But instead of limiting it to 2 elements for specifying the qualified name of the banned property, it should be open ended.
That would allow what @JoshuaKGoldberg requested in #2821 
```json
{"name": ["chai", "assert", "equal"], "message": "Use strictEqual instead of equal"}
```

Any thoughts about adding a wildcard to address @nchen63's comment https://github.com/palantir/tslint/pull/2547#issuecomment-293787648 to ban a property globally?
```js
{"name": ["*", "forEach"], "message": "Use a regular for loop instead"}
// bans [].forEach(...) as well as foo.forEach(...), foo.bar.forEach(...) and foo.bar.baz.forEach(...)
```

**Comment by adidahiya** [comment]
> But instead of limiting it to 2 elements for specifying the qualified name of the banned property, it should be open ended.

sgtm 👍 

> Any thoughts about adding a wildcard

also seems fine, as the user is making it explicit that they are engaging in potentially wild behavior 😂

**Comment by ajafff** [comment]
PR updated to allow banning nested methods as requested in #2821 and added a wildcard for method bans

**Pull request merged**